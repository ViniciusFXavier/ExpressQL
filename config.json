{
  "server": {
    "address": "localhost",
    "port": 3000
  },
  "mongodb": {
    "server": "mongodb://localhost/ExpressQL"
  }
}
