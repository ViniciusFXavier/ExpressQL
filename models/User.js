import mongoose from 'mongoose';
import EncryptController from '../controllers/EncryptController';
import DateController from '../controllers/DateController';

// Defining user Mongoose Schema
const UserSchema = new mongoose.Schema({
  name: String,
  email: { type: String, required: true },
  password: { type: String, required: true },
  birthday: { type: Date },
  admin: { type: Boolean, default: false },
  createdAt: { type: Date, required: true },
  updatedAt: { type: Date, required: true },
});

UserSchema.set('toJSON', { getters: true });

export const User = mongoose.model('User', UserSchema);

/**
 * This function mounts the user data with the right params
 */
const getUserData = (args, update = false) => {
  const name = args.user.name;
  const email = args.user.email;
  const birthday = DateController.isoStringToDate(args.user.birthday);
  const password = EncryptController.encrypt(args.user.password);
  const createdAt = DateController.currentDateISOString();
  const updatedAt = DateController.currentDateISOString();

  const userData = {};

  if (name) userData.name = name;
  if (email) userData.email = email;
  if (birthday) userData.birthday = birthday;
  if (password) userData.password = password;

  if (update) userData.updatedAt = updatedAt;
  else {
    userData.createdAt = createdAt;
    userData.updatedAt = updatedAt;
  }

  return userData;
};

/**
 * This function creates a user on Database
 */
export const createUserDB = args => (
  new Promise((resolve, reject) => {
    // Creating User object that will be added to DB
    const newUser = new User(getUserData(args));

    // Saving new user do DB
    newUser.save((err, res) => (
      err ? reject(err) : resolve(res)
    ));
  })
);

/**
 * This function updates a user on Database
 */
export const updateUserDB = args => (
  new Promise((resolve, reject) => {
    // Updating user
    User.findByIdAndUpdate(args.id, { $set: getUserData(args, true) }, { new: true },
      (err, res) => (
        err ? reject(err) : resolve(res)
      ),
    );
  })
);

/**
 * This functions get a list of users from database
 */
export const getUsersDB = () => (
  new Promise((resolve, reject) => {
    User.find({}).exec((err, res) => (
      err ? reject(err) : resolve(res)
    ));
  })
);

/**
 * This functions get a single user, filtered by it's ID
 */
export const getUserByIdDB = id => (
  new Promise((resolve, reject) => {
    User.find({ _id: id }).exec((err, res) => (
      err ? reject(err) : resolve(res[0])
    ));
  })
);
