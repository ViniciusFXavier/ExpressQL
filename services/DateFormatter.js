/**
 * Gets the current date in ISO format
 * @return {String} The current date string in ISO format
 */
export const getCurrentISOStringDate = () => {
  const date = new Date();

  return date.toISOString();
};

/**
 * Converts a Date to ISO string date
 * @param  {Date} date The date that we wanna convert
 * @return {String}    A date string in ISO format
 */
export const getISOStringFromDate = (date) => {
  if (date instanceof Date) return date.toISOString();

  return null;
};

/**
 * Converts a ISO string date to Date
 * @param  {[type]} isoString A date string in ISO format that we wanna convert
 * @return {[type]}           A date object
 */
export const getDateFromISOString = (isoString) => {
  if (typeof isoString === 'undefined' || isoString === null) {
    return null;
  }

  const date = new Date(isoString);

  return date;
};
