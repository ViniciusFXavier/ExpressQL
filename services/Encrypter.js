/**
 * The objective of this file is to deal with encrypted data,
 * mainly used to generate and check passwords
*/

import bcrypt from 'bcrypt';

// Number of salt rounds
const saltRounds = 10;

/**
 * Encrypts a text, to save as a password
 * @param  {String} text The text that we wanna encrypt
 * @return {String}      Returns the hash (encrypted text)
 */
export const encrypt = (text) => {
  if (typeof text === 'undefined' || text === null) {
    return null;
  }

  const salt = bcrypt.genSaltSync(saltRounds);
  const hash = bcrypt.hashSync(text, salt);

  return hash;
};

/**
 * Compares a text with a hash and check if it's a match.
 * Commonly used to check if a password is valid.
 * @param  {String} text The text that we wanna compare
 * @param  {String} hash The hash we wanna compare
 * @return {Boolean}     Returns true/false. If matches or not
 */

export const compare = (text, hash) => bcrypt.compareSync(text, hash);
