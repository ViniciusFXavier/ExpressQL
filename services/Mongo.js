import mongoose from 'mongoose';

export const connectToMongo = (server) => {
  // Connecting to MongoDB
  console.info(`Connecting to Mongo: ${server}`);
  mongoose.connect(server, {
    useMongoClient: true,
  });
};

export const NONE = 'NONE';
