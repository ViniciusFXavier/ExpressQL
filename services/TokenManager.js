import jwt from 'jsonwebtoken';
import config from '../config.json';

export const encodeToken = (payload) => {
  // Validating entry
  if (typeof payload === 'undefined' || payload === null) {
    return null;
  }
  // Creating token
  const token = jwt.sign(payload, config.jwt.secret);

  return token;
};

export const decodeToken = (token) => {
  // Getting decoded token
  const decoded = jwt.verify(token, config.jwt.secret, (err, dec) => {
    if (err) return null;
    return dec;
  });

  return decoded;
};
