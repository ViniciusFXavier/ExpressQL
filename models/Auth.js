import { compare } from '../services/Encrypter';
import { encodeToken } from '../services/TokenManager';
import { getCurrentISOStringDate } from '../services/DateFormatter';
import { getUsersByFieldDB } from './User';

export const authenticate = args => (
  new Promise((resolve, reject) => {
    const email = args.auth.email;
    const password = args.auth.password;

    // Getting user by email
    getUsersByFieldDB('email', email)
      .then((res) => {
        if (res.length === 0) {
          reject('User not found');
        } else {
          // Defining params
          const user = res[0];
          const hashPassword = user.password;
          // Checking if password if valid
          if (compare(password, hashPassword)) {
            // Creating token object
            const token = encodeToken({
              id: user._id, // eslint-disable-line
              email: user.email,
              createdAt: getCurrentISOStringDate(),
            });
            resolve({ token });
          } else {
            reject('Invalid password');
          }
        }
      })
      .catch(err => reject(err));
  })
);

export const NONE = 'NONE';
