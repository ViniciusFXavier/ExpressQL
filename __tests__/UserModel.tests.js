import { getUsersByFieldDB, getUserByIdDB } from '../models/User';
import { connectToMongo } from '../services/Mongo';
import config from '../config.json';

it('getUsersByField - Must match the snapshot', () => {
  // Connecting to Mongo
  connectToMongo(config.mongodb.server);
  // Defining params
  const email = 'kersul@kersul.com';
  expect.assertions(1);
  return getUsersByFieldDB('email', email).then((data) => {
    // console.info(data);
    expect(data).toMatchSnapshot();
  });
});

it('getUserByIdDB - Must match the snapshot', () => {
  // Connecting to Mongo
  connectToMongo(config.mongodb.server);
  // Defining params
  const id = '59b03e568ce6e41d50459410';
  expect.assertions(1);
  return getUserByIdDB(id).then((data) => {
    // console.info(data);
    expect(data).toMatchSnapshot();
  });
});
