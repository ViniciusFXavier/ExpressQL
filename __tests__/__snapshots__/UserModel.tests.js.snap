// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserByIdDB - Must match the snapshot 1`] = `
Object {
  "__v": 0,
  "_id": "59b03e568ce6e41d50459410",
  "admin": false,
  "birthday": 1990-02-01T10:00:00.000Z,
  "createdAt": 2017-09-06T18:28:38.648Z,
  "email": "kersul@kersul.com",
  "id": "59b03e568ce6e41d50459410",
  "name": "Kersul",
  "password": "$2a$10$mmUyPGv1DJtgLpwG7DULu.tgz0srAxOtrUd3b.fEul9WhG1qTrcce",
  "updatedAt": 2017-09-06T18:28:38.649Z,
}
`;

exports[`getUsersByField - Must match the snapshot 1`] = `
Array [
  Object {
    "__v": 0,
    "_id": "59b03e568ce6e41d50459410",
    "admin": false,
    "birthday": 1990-02-01T10:00:00.000Z,
    "createdAt": 2017-09-06T18:28:38.648Z,
    "email": "kersul@kersul.com",
    "id": "59b03e568ce6e41d50459410",
    "name": "Kersul",
    "password": "$2a$10$mmUyPGv1DJtgLpwG7DULu.tgz0srAxOtrUd3b.fEul9WhG1qTrcce",
    "updatedAt": 2017-09-06T18:28:38.649Z,
  },
]
`;
